File: code/string-algorithms/levenshtein_distance/LevenshteinDistance.java
// Part of Cosmos by OpenGenus Foundation
class LevenshteinDistance {
    private String a, b;
    private int n, m;
    private int[][] dp;

    public LevenshteinDistance(String a, String b) {
        this.a = a;
        this.b = b;
        n = a.length();
        m = b.length();

        dp = new int[n + 1][m + 1];

        initialize();
    }

    private void initialize() {
        for(int i = 0; i <= n; i++) {
            dp[i][0] = i;
        }
        for(int j = 0; j <= m; j++) {
            dp[0][j] = j;
        }
    }

    private int min3(int a, int b, int c) {
        return Math.min(a, Math.min(b, c));
    }

    public int compute() {

        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= m; j++) {

                int c = (a.charAt(i - 1) == b.charAt(j - 1)) ? 0 : 1;

                dp[i][j] = min3(dp[i - 1][j - 1] + c, dp[i - 1][j] + 1, dp[i][j - 1] + 1);
            }
        }
        return dp[n][m];
    }

	public static void main(String[] args) {
	    LevenshteinDistance distance = new LevenshteinDistance("ana", "banana");
        System.out.println(distance.compute());
    }

}

File: code/string_algorithms/src/levenshtein_distance/levenshteindistance.cpp
// Part of Cosmos by OpenGenus Foundation

using namespace std;

int levenshteinDist(string word1, string word2)
{
    int size1 = word1.size();
    int size2 = word2.size();
    vector<vector<int>> dp(size1 + 1, vector<int>(size2 + 1, 0));

    for (int i = 1; i <= size1; i++)
        dp[i][0] = i;

    for (int j = 1; j <= size2; j++)
        dp[0][j] = j;

    for (int i = 1; i <= size1; i++)
    {
        for (int j = 1; j <= size2; j++)
        {
            if (word1[i - 1] == word2[j - 1])
                dp[i][j] = dp[i - 1][j - 1];
            else
                dp[i][j] = min(dp[i - 1][j - 1], min(dp[i][j - 1], dp[i - 1][j])) + 1;
        }
    }

    return dp[size1][size2];
}
